runtime/lua/vim/lsp.lua

```diff
@@ -1239,9 +1243,49 @@ function lsp.buf_request(bufnr, method, params, handler)
   return client_request_ids, _cancel_all_requests
 end
 
---- Sends a request to a server and waits for the response.
+---Sends an async request for all active clients attached to the buffer.
+---Executes the callback on the combined result.
+---Parameters are the same as |vim.lsp.buf_request()| but the return result and callback are
+---different.
+---
+--@param bufnr (number) Buffer handle, or 0 for current.
+--@param method (string) LSP method name
+--@param params (optional, table) Parameters to send to the server
+--@param callback (function) The callback to call when all requests are finished.
+--  Unlike `buf_request`, this will collect all the responses from each server instead of handling them.
+--  A map of client_id:request_result will be provided to the callback
+--
+--@returns (function) A function that will cancel all requests which is the same as the one returned from `buf_request`.
+function lsp.buf_request_all(bufnr, method, params, callback)
+  local request_results = {}
+  local result_count = 0
+  local expected_result_count = 0
+  local cancel, client_request_ids
+
+  local set_expected_result_count = once(function()
+    for _ in pairs(client_request_ids) do
+      expected_result_count = expected_result_count + 1
+    end
+  end)
+
+  local function _sync_handler(err, _, result, client_id)
+    request_results[client_id] = { error = err, result = result }
+    result_count = result_count + 1
+    set_expected_result_count()
+
+    if result_count >= expected_result_count then
+      callback(request_results)
+    end
+  end
+
+  client_request_ids, cancel = lsp.buf_request(bufnr, method, params, _sync_handler)
```

Comment:
Is it always guaranteed that a function which accepts a callback in luv returns before the callback has executed? otherwise client_request_ids is nil when set_expected result_count is called by the _sync_handler callback.

Reply:
Yes I think it should always but I'm not sure

Reply:
If you can confirm that would be great :) I can look into it later too. Otherwise, the PR looks good to me (thanks!). 

I'm guessing you have some goal/utility in mind for lsp.buf_request_all as we don't really use buf_request_sync all that much, and this PR doesn't really affect it. So I'm fine with merging after we confirm no unexpected nils.

Reply:
We will use `buf_request_all` in telescope because `buf_request_sync` is synchronous

Reply:
@tjdevries Do you know re: callback before return result?

Reply:
I asked tj on stream, he says yes but he never proved it